lib/Sema/TypeCheckEffects.cpp

```diff
@@ -3570,16 +3570,40 @@ class CheckEffectsCoverage : public EffectsHandlingWalker<CheckEffectsCoverage>
       if (const AnyTryExpr *tryExpr = dyn_cast<AnyTryExpr>(autoClosure->getSingleExpressionBody()))
         awaitInsertLoc = tryExpr->getSubExpr()->getStartLoc();
     }
+    
+    auto findShorthandOptionalBinding = [&](Expr *anchor) -> VarDecl * {
+      VarDecl *varDecl = nullptr;
+      if (!anchor->getType()->isOptional())
+        return nullptr;
+      anchor->forEachChildExpr([&](Expr *expr) -> Expr * {
+        if (auto *declRef = dyn_cast<DeclRefExpr>(expr)) {
+          if (declRef && declRef->isImplicit()) {
+            varDecl = dyn_cast<VarDecl>(declRef->getDecl());
+            return nullptr;
+          }
+        }
+        return expr;
+      });
+      return varDecl;
+    };
 
     bool downgradeToWarning = llvm::all_of(errors,
         [&](DiagnosticInfo diag) -> bool {
           return diag.downgradeToWarning;
         });
-
-    Ctx.Diags.diagnose(anchor->getStartLoc(), diag::async_expr_without_await)
-      .warnUntilSwiftVersionIf(downgradeToWarning, 6)
-      .fixItInsert(awaitInsertLoc, "await ")
-      .highlight(anchor->getSourceRange());
+    
+    if (auto varDecl = findShorthandOptionalBinding(anchor)) {
+      Ctx.Diags.diagnose(anchor->getStartLoc(), diag::async_expr_without_await)
+        .warnUntilSwiftVersionIf(downgradeToWarning, 6)
+        .fixItInsertAfter(anchor->getStartLoc(),
+                          (" = await " + varDecl->getName().str()).str())
+        .highlight(anchor->getSourceRange());
+    } else {
+      Ctx.Diags.diagnose(anchor->getStartLoc(), diag::async_expr_without_await)
+        .warnUntilSwiftVersionIf(downgradeToWarning, 6)
+        .fixItInsert(awaitInsertLoc, "await ")
+        .highlight(anchor->getSourceRange());
+    }
```

Comment:
Nit: Can this be simplified? 
```suggestion
    auto diag = Ctx.Diags.diagnose(anchor->getStartLoc(), diag::async_expr_without_await)
        .warnUntilSwiftVersionIf(downgradeToWarning, 6)
        .highlight(anchor->getSourceRange());
    if (auto varDecl = findShorthandOptionalBinding(anchor)) {
        diag.fixItInsertAfter(anchor->getStartLoc(),
                          (" = await " + varDecl->getName().str()).str());
    } else {
        diag.fixItInsert(awaitInsertLoc, "await ");
    }
```

Reply:
I like this idea but I tried it and I get the build error: `call to deleted constructor of 'InFlightDiagnostic'`

Reply:
I was able to still simplify it _slightly_ by using `varDecl->getNameStr()`

Reply:
It should be possible to do something like this IIRC ...

Reply:
I managed to find a place in the codebase that does it like in your example, so will try this again.

Reply:
> I like this idea but I tried it and I get the build error: `call to deleted constructor of 'InFlightDiagnostic'`

Because `highlight`, unlike `diagnose`, returns a reference, and the `auto` in `auto diag` gets deduced as a non-reference, which implies a copy or a call to the deleted copy constructor. `auto &diag` should work.

Reply:
Ah I see, thanks!